winrm/winrm.go

```diff
@@ -0,0 +1,271 @@
+	// - pass only client cert and key
+	// - 30 sec timeout
+	if ca := CACert(); ca != nil {
+		endpoint = winrm.NewEndpoint(host, httpsPort, true, false, ca, ClientCert(), ClientKey(), 30*time.Second)
+	} else {
+		endpoint = winrm.NewEndpoint(host, httpsPort, true, true, nil, ClientCert(), ClientKey(), 30*time.Second)
+		logger.Warningf("Skipping CA server verification, using the --insecure option")
+	}
+	logger.Debugf("Creating WinRM secure clinet connection on %s %d", host, httpsPort)
+	// if we don't have defined user
+	// use the default administrator on windows machine
+	if user == "" {
+		user = defaultWinndowsUser
+	}
+
+	cli.conn, cli.err = winrm.NewClient(endpoint, user, "")
+	if cli.err != nil {
+		cli.err = errors.Annotatef(cli.err, "Connot create WinRM secure client conn")
+	}
+	return cli
+}
+
+// GetPasswd callback for different semantics that a client
+// could use for secure authentication
+type GetPasswd func() (string, error)
+
+// TTYGetPasswd will be valid if it's used from a valid TTY input,
+// This can be used in conjunction with UseClient
+func TTYGetPasswd() (string, error) {
+	// very important that os.Stdin.Fd() needs to be a valid tty.
+	// if the file descriptor dosen't point to a terminal this will fail
+	// with the message inappropriate ioctl for device
+	pass, err := terminal.ReadPassword(int(os.Stdin.Fd()))
+	if err != nil {
+		return "", errors.Annotatef(err, "Can't retrive password from terminal")
+	}
+	return string(pass), nil
+}
+
+// NewClient binds new winrm client for initiating connections with the winrm listener
+// Note: use this function before winrm.LoadClientKeys()
+func NewClient(user, host string, gp GetPasswd, stdout io.Writer) *Client {
+	cli := &Client{}
+	winrm.DefaultParameters.TransportDecorator = nil
+
+	var endpoint *winrm.Endpoint
+	if ca := CACert(); ca != nil {
+		endpoint = winrm.NewEndpoint(host, httpPort, false, false, ca, nil, nil, 30*time.Second)
+	} else {
+		endpoint = winrm.NewEndpoint(host, httpPort, false, true, nil, nil, nil, 30*time.Second)
+		logger.Warningf("Skipping CA server verification, using the --insecure option")
+	}
+
+	logger.Debugf("Creating WinRM clinet connection on %s port %d", host, httpPort)
+
+	// if we don't have defined user
+	// use the default administrator on windows machine
+	if user == "" {
+		user = defaultWinndowsUser
+	}
+
+	if gp != nil { // get password from a given way
+		fmt.Fprintf(stdout, "[Winrm] >> ")
+		cli.pass, cli.err = gp()
+		if cli.err != nil {
+			return cli
+		}
+		fmt.Fprintf(stdout, "\n")
+	}
+
+	cli.conn, cli.err = winrm.NewClient(endpoint, user, cli.pass)
+	if cli.err != nil {
+		cli.err = errors.Annotatef(cli.err, "Connot create WinRM client conn")
+	}
+
+	return cli
+}
+
+// Run command and direct output to stdout and errors to stderr
+// If the Run successfully executes the cmd it returns nil on call of Error()
+func (c *Client) Run(command string, stdout io.Writer, stderr io.Writer) {
```

Comment:
@hoenirvili sorry, I was confused because there's a bit of interchanging of "Client" and "Connection" in this thread. Connection implies, to me, something that needs to be closed. The client isn't connection-oriented, though, so Close isn't necessary. So the API would just be:

```go
package winrm

func NewClient(Config) *Client

type Client struct {...}

func (*Client) Ping() error
func (*Client) Run(cmd string, stdout, stderr io.Writer) error
```

My previous statement about there being no sensible place for a single CA cert for all servers remains, so I still don't think it's sensible to have a global client. Not to mention it's racy: if you had two things in the same process trying to manually provision two different machines, we'd be in trouble. This isn't just academic, as we may want to move the bulk of manual provisioning responsibility to the controller sooner or later.

Seeing as you're passing in "DefaultClient" and "SecureClient" *anyway*, I see no reason why you would want to then pass that into a global and then use the global. Why not use the values directly? Doing so avoids concurrency issues, and makes testing easier.

Finally, on usage in PR juju/juju#6523. As above I was mistaken about the need for Close, so you needn't create a new Client per Run. Instead, just pass the pair of clients in (or a Client constructor function), and use them directly for the lifetime of the ProvisionMachine call. The Ping calls would be more idiomatic if the method returned an error, e.g.

at https://github.com/juju/juju/pull/6523/files#diff-8eee7a4e81e6a28a0cbc14fa50ae0b6bR128

```go
winrm.Ping()
first := winrm.Error()
if first == nil {
```

would become

```go
first := args.SecureClient.Ping()
if first == nil {
```

The Run calls would be slightly more verbose (i.e. with immediate error checking), but (a) there aren't that many Run calls, and (b) the errors will be directly relatable to the source (rather than "an error happened somewhere along the way", which you may or may not be able to determine the context of from the error message).

In case all of that is unclear, here is what I think InitAdministratorUser should look like:

```go
func InitAdministratorUser(args manual.ProvisionMachineArgs) error {
        logger.Infof("Initalising %q, user %q", args.Host, args.User)
        logger.Infof("Using WinRM secure client as user %q on %q", args.Host, args.User)
 
        first := args.SecureClient.Ping()
        if first == nil {
                logger.Infof("WinRM secure authentication is already enabled on the host %q", args.Host)
                return nil
        }
        logger.Debugf("Secure authentication is not enabled on user %q", args.User)
 
        if args.DefaultClient == nil { // if it's not already custom set use the default
                // at this state the secure listener is not enabled and
                // we must enable it first
                args.DefaultClient = winrm.NewClient(args.User, args.Host, winrm.TTYGetPasswd, args.Stdout)
        }

        last := args.DefaultClient.Ping()
        if last == nil {
                logger.Infof("WinRM normal connection is enabled on the host %q", args.Host)
                return nil
        }
        logger.Errorf("Normal authentication is not enabled on the host %q", args.Host)
 
        if first != nil && last != nil {
                return errors.Annotatef(first, last.Error())
        }
 
        if first != nil {
                return first
        }
 
        return last
}
```

(Bear in mind that ProvisionMachineArgs is a value type, so the update to DefaultClient won't persist beyond the end of the function. I think it might be better to require that ProvisionMachineArgs has a SecureClient and a DefaultClient on entry to the exported functions, using a Validate method.)